include/wil/safearrays.h

```diff
@@ -0,0 +1,1084 @@
+
+        inline void __stdcall SafeArrayDestroy(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayDestroy(psa));
+        }
+
+        inline void __stdcall SafeArrayLock(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayLock(psa));
+        }
+
+        inline void __stdcall SafeArrayUnlock(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayUnlock(psa));
+        }
+
+        template<typename T>
+        inline void __stdcall SafeArrayAccessData(SAFEARRAY* psa, T*& p) WI_NOEXCEPT
+        {
+            WI_ASSERT(sizeof(T) == ::SafeArrayGetElemsize(psa));
+            FAIL_FAST_IF_FAILED(::SafeArrayAccessData(psa, reinterpret_cast<void**>(&p)));
+        }
+
+        inline void __stdcall SafeArrayUnaccessData(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            FAIL_FAST_IF_FAILED(::SafeArrayUnaccessData(psa));
+        }
+
+        inline VARTYPE __stdcall SafeArrayGetVartype(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            VARTYPE vt = VT_NULL;   // Invalid for SAs, so use to mean SA was null
+            if ((psa != nullptr) && FAILED(::SafeArrayGetVartype(psa, &vt)))
+            {
+                vt = VT_EMPTY;  // Invalid for SAs, use to mean type couldn't be determined
+            }
+            return vt;
+        }
+
+        inline ULONG __stdcall SafeArrayGetLockCount(SAFEARRAY* psa) WI_NOEXCEPT
+        {
+            return (psa != nullptr) ? psa->cLocks : 0U;
+        }
+
+        inline HRESULT __stdcall SafeArrayCreate(VARTYPE vt, UINT dims, SAFEARRAYBOUND* sab, SAFEARRAY*& psa) WI_NOEXCEPT
+        {
+            WI_ASSERT(dims > 0);
+            psa = ::SafeArrayCreate(vt, dims, sab);
+            RETURN_LAST_ERROR_IF_NULL(psa);
+            WI_ASSERT(vt == details::SafeArrayGetVartype(psa));
+            return S_OK;
+        }
+
+        inline HRESULT __stdcall SafeArrayGetDimSize(SAFEARRAY* psa, UINT dims, ULONG* result) WI_NOEXCEPT
+        {
+            RETURN_HR_IF(E_INVALIDARG, ((dims == 0) || (dims > psa->cDims)));
+            // Read from the back of the array because SAFEARRAYs are stored with the dimensions in reverse order
+            *result = psa->rgsabound[psa->cDims - dims].cElements;
+            return S_OK;
+        }
+
+        inline HRESULT __stdcall SafeArrayGetTotalSize(SAFEARRAY* psa, ULONG* result) WI_NOEXCEPT
+        {
+            if (psa != nullptr)
+            {
+                ULONGLONG elements = 1;
+                for (UINT i = 0; i < psa->cDims; ++i)
+                {
+                    // Determine the total size of the safearray by multiplying
+                    // the sizes of each dimension together
+                    elements *= psa->rgsabound[i].cElements;
+                    if (elements > ULONG_MAX)
+                    {
+                        RETURN_HR(HRESULT_FROM_WIN32(ERROR_ARITHMETIC_OVERFLOW));

```

Comment:
This seems kind of excessive and seems to be the only thing stopping something as simple as `size` from returning its result (as opposed to being an out param). Why not just make it a `size_t`?

Reply:
I really wanted to do that, but `size_t` is platform dependent, so we can't assume the size, but it will be 32-bit on some systems, which won't work.  I could define a `size_type` that is always 64-bit, but that solve anything because even a 64-bit value could overflow when multiplying an arbitrary number of 32-bit values.

Reply:
I assume that all of the memory for `SAFEARRAY`s must be in virtual memory, no? If that's the case, it's impossible for the total size to exceed `size_t` unless the data is corrupt

Reply:
I'm not sure that's true for 32-bit compilations.

Reply:
The reason is SAFEARRAY uses ULONG for a size.  So, when computing the total size, we have to make sure we don't overflow a ULONG, so I use a ULONGLONG, check for overflow again ULONG_MAX, and then downcast to a ULONG after confirming it's safe.

Reply:
The `SAFEARRAY` type seems to use `ULONG` to store the size of each dimension, however I don't see anything in the documentation that would indicate that the product of dimensions can't exceed the bounds of `ULONG`. So, I checked the definition of `SafeArrayCreate` and it appears to use `ULONG` to calculate the size to allocate (despite `IMalloc` accepting a `size_t` argument...).

So `ULONG` is probably fine here, although theoretically the API could be updated in the future to allow for larger allocations. One piece of feedback is to consider using the existing safe math functions that exist in `intsafe.h` instead of rolling your own, but otherwise I concede here.